### DISCUSSION TITLE
Are fastjet-contrib modules in project scope?

### DISCUSSION THREAD

**Discussion by matthewfeickert** [scikit-hep/fastjet#90] [Ideas]
I'm jumping the gun a bit here, but given [this Tweet today](https://twitter.com/suchi_kulkarni/status/1541485635231514626?s=20&t=ehdpYKAuIq-HPNdSN9-NXQ) (2022-06-27) from @suchitakulkarni it seems that once there is a stable `v3.3.4`[^1] out [on PyPI](https://pypi.org/project/fastjet/3.3.4.0rc9/#history) that adding support for the [FastJet Contrib modules](https://fastjet.hepforge.org/contrib/contents/latest.html) would be of general interest. The level of work there would probably be another full IRIS-HEP Fellow project though.

Is this in project scope? Or is this too expansive and would this be better suited as another library that would be tightly constrained in version to `fasjet` in a similar way to how [`hist`'s version is tightly scoped to `boost-histogram`](https://iscinumpy.dev/post/bound-version-constraints/#examples-of-acceptable-caps)?

Tagging other people from Twitter who might be interested in this development: @kratsg, @mhance, @mattleblanc, @OzAmram.

[^1]: Maybe I don't understand the versioning convention here and it actually _is_ `v3.3.4.0`?

**Comment by matthewfeickert** [comment]
@jpivarski @aryan26roy @henryiii have any thoughts here?

**Reply by jpivarski** [reply]
We were waiting for the fastjet-contrib modules to move from svn to git so that we can make them a git submodule. I don't know if that has happened yet.

**Reply by matthewfeickert** [reply]
@gavinsalam Can you comment here on what the status is to move to Git? (Like the main FastJet package did with its migration to GitLab https://gitlab.com/fastjet/fastjet)

**Reply by gavinsalam** [reply]
It is still in our plans, but we don't currently have a timeline for it to happen. Sorry not to be more helpful. 

A Python interface would itself also involve some non-negligible amount of work, and there is a question of who would maintain it -- is this something you (scikit-hep/fastjet) would envisage doing? FJContrib itself simply collects 3rd party contributed tools (with quite a diverse range of interfaces), and usually does not seek to augment them directly.

**Reply by matthewfeickert** [reply]
> It is still in our plans, but we don't currently have a timeline for it to happen. Sorry not to be more helpful.

@gavinsalam Thanks for this info still &mdash; just knowing that is in scope at all is helpful. Is there a place that interested parties can watch for updates on timelines or project scope plans?

> A Python interface would itself also involve some non-negligible amount of work, and there is a question of who would maintain it -- is this something you (scikit-hep/fastjet) would envisage doing? FJContrib itself simply collects 3rd party contributed tools (with quite a diverse range of interfaces), and usually does not seek to augment them directly.

Those are good questions and part of the reason I opened up this GitHub Discussion to better understand what is in scope. I think that if this was considered in scope by all parties we would need to get a [good IRIS-HEP Fellow](https://iris-hep.org/fellows.html) paid to work on this as I don't think that anyone tagged in this Discussion so far has the time or bandwidth to do the development.

@gavinsalam maybe to guide this discussion a bit, as I am far from an expert user of [`fastjet-contrib`](https://fastjet.hepforge.org/contrib/), can you comment on how 3rd party extensions are selected to be added to `fastjet-contrib` and once they are added who has the responsibility to maintain them? There is no current agreed upon Python bindings or API for `fastjet-contrib` either, correct? (So "non-negligible amount of work" is pretty spot on :slightly_smiling_face:)

**Reply by gavinsalam** [reply]
We usually announce significant updates on https://fastjet.fr/ . 

fastjet-contrib is very open: anyone who would like their (C++) contribution to be added simply needs to follow some basic guidelines for the package structure & tests and then ask us for access to the repo. Typically, contributors remain responsible for fixing issues in their specific contribution, while the FastJet authors usually manage the framework.

Curently, there are indeed no agreed upon Python bindings or API for fastjet-contrib; one of the complications is that there is not all that much uniformity from one contrib to the next, because each one is often aiming to solve a different class of problem.

**Comment by jmduarte** [comment]
Just to add to this discussion, CMS maintains its own mirror of `fj-contrib` on GitHub. 

Latest version: https://github.com/cms-externals/fastjet-contrib/tree/cms/v1.044

**Reply by jmduarte** [reply]
Update: `fastjet-contrib` is now merged as a module! https://github.com/scikit-hep/fastjet/pull/143
We should probably create a roadmap with some community/user input on which modules are the most useful to add interfaces to (e.g. SoftDrop, Nsubjettiness, EnergyCorrelationFunctions, etc.)

**Reply by lgray** [reply]
I recommend we develop that list of interesting modules and then mark the integration of all of them as the next release.

Given the general and steady community interest in jet algorithms I think it'll come together easily!

**Reply by ek234** [reply]
@jmduarte 
@lgray 
Is this assigned to anyone? If not, I can try adding interfaces to a few modules. What modules would you like to have?
